### ISSUE TITLE
process cleanup for ReactorBuilder-based tests loops indefinitely if processes do not exit

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3398]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)| @zooko reported|
|-|-|
|Trac ID|trac#3398|
|Type|defect|
|Created|2008-08-27 00:25:27Z|
|Branch|https://github.com/twisted/twisted/tree/process-cleanup-loop-3398-2|

http://buildbot.twistedmatrix.com/builders/cygwin-py2.5-select/builds/1

Produces many hundreds or thousands of this line in the test.log:

2008-08-26 15:29:02-0700 [-] ReactorBuilder.tearDown reaping some processes {2540: <PTYProcess pid=2540 status=-1>}

Followed by many of this:

2008-08-26 15:29:10-0700 [-] ReactorBuilder.tearDown reaping some processes {2524: <PTYProcess pid=2524 status=-1>}

and eventually I stopped it.

<details><summary>Searchable metadata</summary>

```
trac-id__3398 3398
type__defect defect
reporter__zooko zooko
priority__normal normal
milestone__ 
branch__branches_process_cleanup_loop_3398_2 branches/process-cleanup-loop-3398-2
branch_author__itamarst__exarkun itamarst, exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1219796727000000 1219796727000000
changetime__1333934582000000 1333934582000000
version__None None
owner__itamar itamar
cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

It seems the process started by `twisted.internet.test.test_process.ProcessTestsBuilder_PollReactor.test_childConnectionLost` never exits.  It is supposed to exit when it receives a blank line on stdin, which the test sends it after sending two other lines.

It seems like the test method actually ran to completion and all the events it expected did happen because the infinite log message happens in `tearDown` which should only run after the test method has returned.

So the question here may be about why the child process isn't exiting when it is expected to.

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_2"></a>|@zooko commented|
|-|-|

Maybe it is waiting for a DOS-style blank line: '\r\n' ?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

Maybe.  I don't have cygwin installed anywhere, so I can't test this.

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_4"></a>|@zooko set status to `closed`|
|-|-|

duplicate of [#3529](https://github.com/twisted/twisted/issues/3529)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set status to `reopened`|
|-|-|

Even if cygwin is broken, the tests shouldn't get stuck in a loop.  [#3529](https://github.com/twisted/twisted/issues/3529) can be for fixing cygwin support somehow.  I'm using this ticket to fix the getting stuck behavior.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [25347]) Branching to 'process-cleanup-loop-3398'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set owner to @zooko|
|-|-|

Can you try the branch on cygwin?

I expect the tests will fail, but the suite should run to completion at least.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst set owner to @itamarst|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_10"></a>|@itamarst commented|
|-|-|

This is happening on non-cygwin platforms as well, e.g.:
http://buildbot.twistedmatrix.com/builders/fedora32-py2.5-reactors/builds/1553/steps/poll/logs/test.log

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_11"></a>|@itamarst commented|
|-|-|

(In [34151]) Branching to 'process-cleanup-loop-3398-2'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_12"></a>|@itamarst set status to `closed`|
|-|-|

(In [34153]) Merge process-cleanup-loop-3398-2: Cleanup after zombie test processes.

Author: exarkun
Review: itamar
Fixes: [#3398](https://github.com/twisted/twisted/issues/3398)

Hopefully this will make the Fedora buildbot stop getting into infinite loops.
